src/mds/Server.cc

```diff
@@ -7929,6 +7929,23 @@ void Server::handle_client_unlink(MDRequestRef& mdr)
   if (!dn)
     return;
 
+  CDentry::linkage_t *dnl = dn->get_linkage(client, mdr);
+  ceph_assert(!dnl->is_null());
+  CInode *in = dnl->get_inode();
+
+  // wait the previous unlink finshes thoroughly.
+  CDentry* _dn = in->get_projected_parent_dn();
+  if (_dn->get_dir()->inode->is_stray()) {
```

Comment:
So I have a little time to comment on this tonight. Do you expect that `rdlock_path_xlock_dentry` will trigger a stray reintegration (I cannot recall off-hand and the code is spaghetti).  I would be clear that we're not waiting for "unlink finishes thoroughly" but that even acquiring the necessary locks on the dentry will automatically cause the rename to be fired off, which we should wait for.

If that rename is **not** fired off by `rdlock_path_xlock_dentry`, I would be concerned this unlink could wait forever if nothing else touches that dentry to trigger reintegration.